Wqcoprocess

David-Simonetti-ND opened the pull request:
Initial PR for issue #2825

David-Simonetti-ND commented:
I am unhappy with how the end_coprocess code is right now because it is very much a copy paste of the code work_queue_factory uses. would there be some way to unify these functions?

btovar commented:
@David-Simonetti-ND, yes, that's a good idea.  Could you add such a function to `dttools/src/process.*` with a name such as `process_kill_waitpid(pid, timeout)` ?  It should work in the same way as `process_waitpid`.

btovar commented on dttools/src/process.c:
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
+	int status; 
+	int ret_val = 1; // return value from this function: 0 means process exited cleanly, 1 means there was an error
+	while (kill(pid, current_signal) == 0) // send signal to process
+	{
```
to match the style of the other functions, move this brace to previous line

btovar commented on dttools/src/process.c:
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
+	int ret_val = 1; // return value from this function: 0 means process exited cleanly, 1 means there was an error
+	while (kill(pid, current_signal) == 0) // send signal to process
+	{
+		if (timeout != 0) {alarm(timeout);} // and set an alarm for timeout seconds
```
expand to three lines ( if {..., alarm,   } )

btovar commented on dttools/src/process.c:
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
+	{
+		if (timeout != 0) {alarm(timeout);} // and set an alarm for timeout seconds
+		num_attempts++;
+		if (num_attempts == max_attempts) // change to sending SIGKILL after max_attempts tries
```
move brace to previous line, same for all the rest of the function

btovar replied:
Also, `if(num_attempts >= max_attempts)`

btovar commented on dttools/src/process.c:
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
+int process_kill_waitpid(pid_t pid, int timeout, int max_attempts)
+{
+	int flags = 0;
+        struct sigaction new_action, old_action;
```
Is the file function indented uniformly? The rest of the file uses only tabs.

btovar commented on dttools/src/process.c:
```diff
@@ -98,6 +98,66 @@ struct process_info *process_waitpid( pid_t pid, int timeout)
+				break;
+			}
+		}
+		else if (return_pid == 0) // if the return value is 0, it means we set the timeout to zero and nothing changed, so we try to signal and wait for the process again
```
Doesn't this trigger a busy-wait? I.e.. If the timeout is zero, and waitpid is no hang, then if the processes has not changed, this loop will be executed over and over again with no pause in between.

btovar replied:
I think it should be ok to always require `process_kill_waitpid` to have a positive timeout and use waitpid without the no hang flag. Also, I'm not sure about the re-sending of signals until max_attempts. Shouldn't we just send the signal once and then do the sigkill?  What worries me is that a processes may modify its cleanup when receiving the same signal in succession (e.g. first signal triggers cleanup, second or more terminates immediately).

David-Simonetti-ND commented:
I reworked process_kill_waitpid to use process_waitpid because there may be cases where a given child has already return and is in the completed process list. It now only sends SIGTERM once, waits for timeout seconds, and then sends SIGKILL and attempts to wait for another timeout seconds. At the end it just gives up.
I also added functionality in the work_queue_worker for checking if the coprocess specified in the command line has stopped running and restarting it if this is the case.

btovar commented:
Looks good. Ready to merge?
